src/v/cluster/rm_stm.cc

```diff
@@ -1203,20 +1203,31 @@ ss::future<result<kafka_result>> rm_stm::idempotent_replicate(
   model::record_batch_reader br,
   raft::replicate_options opts,
   ss::lw_shared_ptr<available_promise<>> enqueued) {
-    auto producer = maybe_create_producer(bid.pid);
-    co_return co_await producer
-      ->run_with_lock([&](ssx::semaphore_units units) {
-          return do_sync_and_idempotent_replicate(
-            producer,
-            bid,
-            std::move(br),
-            opts,
-            std::move(enqueued),
-            std::move(units));
-      })
-      .finally([this, producer] {
-          _producer_state_manager.local().touch(*producer, _vcluster_id);
-      });
+    try {
+        auto producer = maybe_create_producer(bid.pid);
+        co_return co_await producer
+          ->run_with_lock([&](ssx::semaphore_units units) {
+              return do_sync_and_idempotent_replicate(
+                producer,
+                bid,
+                std::move(br),
+                opts,
+                std::move(enqueued),
+                std::move(units));
+          })
+          .finally([this, producer] {
+              _producer_state_manager.local().touch(*producer, _vcluster_id);
+          });
+    } catch (const cache_full_error& e) {
```

Comment:
should this go inside maybe_create_producer()? That is used in multiple places, but that _also_ is a problem I think because if that happens in the state machine apply logic (producer evicted between applying to the log and applying to the stm).